x/auth/keeper/migrations.go

```diff
@@ -0,0 +1,146 @@
+package keeper
+
+import (
+	"fmt"
+
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+
+	"github.com/gogo/protobuf/grpc"
+	"github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+const (
+	delegatorDelegationPath = "/cosmos.staking.v1beta1.Query/DelegatorDelegations"
+	balancesPath            = "/cosmos.bank.v1beta1.Query/AllBalances"
+)
+
+// Migrator is a struct for handling in-place store migrations.
+type Migrator struct {
+	keeper      AccountKeeper
+	queryServer grpc.Server
+}
+
+// NewMigrator returns a new Migrator.
+func NewMigrator(keeper AccountKeeper, queryServer grpc.Server) Migrator {
+	return Migrator{keeper: keeper, queryServer: queryServer}
+}
+
+// Migrate1to2 migrates from version 1 to 2.
+func (m Migrator) Migrate1to2(ctx sdk.Context) error {
+	m.keeper.IterateAccounts(ctx, func(account types.AccountI) (stop bool) {
+		asVesting := vesting(account)
+		if asVesting == nil {
+			return false
+		}
+
+		addr := account.GetAddress().String()
+		balance := getBalance(
+			ctx,
+			addr,
+			m.queryServer,
+		)
+
+		delegations := getDelegatorDelegations(
+			ctx,
+			addr,
+			m.queryServer,
+		)
+
+		if delegations.IsAllGTE(asVesting.GetOriginalVesting()) {
+			delegations = asVesting.GetOriginalVesting()
+		}
+
+		if balance.IsAllLT(delegations) {
+			balance = balance.Add(delegations...)
+		}
+
+		asVesting.TrackDelegation(ctx.BlockTime(), balance, delegations)
```

Comment:
I'm having a bit of trouble wrapping my head around this. Can you provide an explanation? 

We're taking an account's balance and adding the delegation sum so as to treat the balances as if the account had never delegated to begin with, then track delegations as if it was done in one fell swoop? Is that correct?

Reply:
It is correct.

We think this is a correct way to account for `DelegatedFree` and `DelegatedVesting` fields because in the end we're just accounting what how much has been delegated.

It's like the user is doing a big delegation in the first place.